src/js/fs.js

```diff
@@ -382,6 +382,196 @@ fs.readdirSync = function(path) {
+
+    Readable.call(this, {defaultEncoding: options.encoding});
+
+    this.bytesRead = 0;
+    this.path = path;
+    this._autoClose = options.autoClose;
+    this._bufferSize = options.bufferSize;
+    this._fd = options.fd;
+    this._buff = new Buffer(options.bufferSize);
+
+    var self = this;
+    if (util.isNullOrUndefined(this._fd)) {
+      fs.open(self.path, options.flags, options.mode, function(err, _fd) {
+        if (err) {
+          throw err;
+        }
+        self._fd = _fd;
+        self.emit('open', self._fd);
+        self.doRead();
+      });
+    }
+
+    this.on('open', function(/* _fd */) {
+      this.emit('ready');
+    });
+
+    if (this._autoClose) {
+      this.on('end', function() {
+        closeFile(self);
+      });
+    }
+  }
+
+
+  util.inherits(ReadStream, Readable);
+
+
+  ReadStream.prototype.doRead = function() {
+    var self = this;
+    fs.read(this._fd, this._buff, 0, this._buff.length, null,
+            function(err, bytes_read/* , buffer*/) {
+      if (err) {
+        if (self._autoClose) {
+          closeFile(self);
+        }
+        throw err;
+      }
+
+      self.bytesRead += bytes_read;
+      if (bytes_read === 0) {
+        // Reached end of file.
+        // null must be pushed so the 'end' event will be emitted.
+        self.push(null);
+      } else {
+        self.push(bytes_read == self._buff.length ?
+                  self._buff : self._buff.slice(0, bytes_read));
+        self.doRead();
+      }
+    });
+  };
+
+
+  fs.createReadStream = function(path, options) {
+    return new ReadStream(path, options);
+  };
+
+
+  function WriteStream(path, options) {
+    if (!(this instanceof WriteStream)) {
+      return new WriteStream(path, options);
+    }
+
+    var defaultOptions = {
+      flags: 'w',
+      fd: null,
+      mode: 438, // == octal 0666
+      autoClose: true,
+    };
+
+    options = adjust_options(options, defaultOptions);
+
+    Writable.call(this);
+
+    this._fd = options._fd;
+    this._autoClose = options.autoClose;
+    this.bytesWritten = 0;
+
+    var self = this;
+    if (!this._fd) {
+      fs.open(path, options.flags, function(err, _fd) {
+        if (err) {
+          throw err;
+        }
+        self._fd = _fd;
+        self.emit('open', self._fd);
+      });
+    }
+
+    this.on('open', function(/* _fd */) {
+      self.emit('ready');
+    });
+
+    if (this._autoClose) {
+      this.on('finish', function() {
+        closeFile(self);
+      });
+    }
+  }
+
+
+  util.inherits(WriteStream, Writable);
+
+
+  WriteStream.prototype._write = function(chunk, callback, onwrite) {
+    var buff = new Buffer(chunk);
+    var self = this;
+    fs.write(this._fd, buff, 0, buff.length,
+             function(err, bytes_written/* , buffer */) {
+      if (err) {
+        if (self._autoClose) {
+          closeFile(self);
+        }
+        throw err;
+      }
+      this.bytesWritten += bytes_written;
+    });
+    if (callback) {
```

Comment:
So we have two callbacks? `callback` and `onwrite`? Should these be called inside the `fs.write`?